From 0a9c868683a67eebf3c6db6b14236c6e972f8bd5 Mon Sep 17 00:00:00 2001
From: Ding Yunliang <yunliang.ding@intel.com>
Date: Mon, 2 Jul 2018 15:12:59 +0800
Subject: [PATCH 1983/2313] media: intel-ipu4: Fix debugfs dir removal issue

Message for Open Source:
Issue happened when rmmod ipu module, as IPU debugfs will recursively
removed first, then isys and psys debugfs will be removed without
parent, kernel crash happened.

[ 4727.399285] BUG: unable to handle kernel NULL pointer dereference
at 00000000000000a8
[ 4727.406666] IP: down_write+0x16/0x40
[ 4727.409213] PGD 0 P4D 0
[ 4727.536109] task: ffff91e5756aa4c0 task.stack: ffffb2ac407cc000
[ 4727.541296] RIP: 0010:down_write+0x16/0x40
[ 4727.544430] RSP: 0018:ffffb2ac407cfb68 EFLAGS: 00010296
[ 4727.600766] Call Trace:
[ 4727.602054]  debugfs_remove_recursive+0x50/0x1b0
[ 4727.605780]  ipu_psys_remove+0x38/0x240 [intel_ipu4_psys]
[ 4727.610383]  ipu_bus_remove+0x25/0x40 [intel_ipu4]
[ 4727.614304]  device_release_driver_internal+0x18a/0x250
[ 4727.618708]  device_release_driver+0x12/0x20
[ 4727.622035]  bus_remove_device+0xf0/0x160
[ 4727.625075]  device_del+0x1d9/0x350
[ 4727.627528]  ? synchronize_srcu_expedited+0x27/0x30

Message for Internal:
[Issue/Feature] Same as above.
[Root Cause/Changes] Same as above.

Change-Id: I17fc312971b6e9dc4c7d31fb15f4394455da1c24
Signed-off-by: Ding Yunliang <yunliang.ding@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 drivers/media/pci/intel/ipu-isys.c | 3 ++-
 drivers/media/pci/intel/ipu-psys.c | 3 ++-
 drivers/media/pci/intel/ipu.c      | 5 +++++
 3 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/media/pci/intel/ipu-isys.c b/drivers/media/pci/intel/ipu-isys.c
index c757534..5e4aecd 100644
--- a/drivers/media/pci/intel/ipu-isys.c
+++ b/drivers/media/pci/intel/ipu-isys.c
@@ -923,7 +923,8 @@ static void isys_remove(struct ipu_bus_device *adev)
 	struct isys_fw_msgs *fwmsg, *safe;
 
 	dev_info(&adev->dev, "removed\n");
-	debugfs_remove_recursive(isys->debugfsdir);
+	if (isp->ipu_dir)
+		debugfs_remove_recursive(isys->debugfsdir);
 
 	list_for_each_entry_safe(fwmsg, safe, &isys->framebuflist, head) {
 		dma_free_attrs(&adev->dev, sizeof(struct isys_fw_msgs),
diff --git a/drivers/media/pci/intel/ipu-psys.c b/drivers/media/pci/intel/ipu-psys.c
index 5c550cc..6dc2821 100644
--- a/drivers/media/pci/intel/ipu-psys.c
+++ b/drivers/media/pci/intel/ipu-psys.c
@@ -1621,7 +1621,8 @@ static void ipu_psys_remove(struct ipu_bus_device *adev)
 	struct ipu_psys *psys = ipu_bus_get_drvdata(adev);
 	struct ipu_psys_pg *kpg, *kpg0;
 
-	debugfs_remove_recursive(psys->debugfsdir);
+	if (isp->ipu_dir)
+		debugfs_remove_recursive(psys->debugfsdir);
 
 	flush_workqueue(IPU_PSYS_WORK_QUEUE);
 
diff --git a/drivers/media/pci/intel/ipu.c b/drivers/media/pci/intel/ipu.c
index f460231..d1f955e 100644
--- a/drivers/media/pci/intel/ipu.c
+++ b/drivers/media/pci/intel/ipu.c
@@ -272,7 +272,12 @@ static int ipu_init_debugfs(struct ipu_device *isp)
 
 static void ipu_remove_debugfs(struct ipu_device *isp)
 {
+	/*
+	 * Since isys and psys debugfs dir will be created under ipu root dir,
+	 * mark its dentry to NULL to avoid duplicate removal.
+	 */
 	debugfs_remove_recursive(isp->ipu_dir);
+	isp->ipu_dir = NULL;
 }
 
 static int ipu_pci_config_setup(struct pci_dev *dev)
-- 
2.7.4

